cosmos/operators/virtualenv.py

```diff
@@ -101,11 +113,113 @@ def run_subprocess(self, command: list[str], env: dict[str, str], cwd: str) -> F
+        if self.is_virtualenv_dir_temporary and self.virtualenv_dir and self.virtualenv_dir.exists():
+            self.log.info(f"Deleting the Python virtualenv {self.virtualenv_dir}")
+            shutil.rmtree(str(self.virtualenv_dir), ignore_errors=True)
+
     def execute(self, context: Context) -> None:
-        output = super().execute(context)
-        if self._venv_tmp_dir:
-            self._venv_tmp_dir.cleanup()
-        logger.info(output)
+        try:
+            output = super().execute(context)
+        except Exception:
+            self.log.error(traceback.format_exc())
+        else:
+            self.log.info(output)
+        finally:
+            self.clean_dir_if_temporary()
+
+    def on_kill(self) -> None:
+        self.clean_dir_if_temporary()
+
+    def prepare_virtualenv(self) -> str:
+        self.log.info(f"Creating or updating the virtualenv at `{self.virtualenv_dir}")
+        py_bin = prepare_virtualenv(
+            venv_directory=str(self.virtualenv_dir),
+            python_bin=PY_INTERPRETER,
+            system_site_packages=self.py_system_site_packages,
+            requirements=self.py_requirements,
+            pip_install_options=self.pip_install_options,
+        )
+        return py_bin
+
+    def _get_or_create_venv_py_interpreter(self) -> str:
+        """Helper method that parses virtual env configuration
+        and returns a DBT binary within the resulting virtualenv"""
+
+        # No virtualenv_dir set, so create a temporary virtualenv
+        if self.virtualenv_dir is None or self.is_virtualenv_dir_temporary:
+            self.log.info("Creating temporary virtualenv")
+            with TemporaryDirectory(prefix="cosmos-venv") as tempdir:
+                self.virtualenv_dir = Path(tempdir)
+                py_bin = self.prepare_virtualenv()
+            return py_bin
+
+        # Use a reusable virtualenv
+        self.log.info(f"Checking if the virtualenv lock {str(self.__lock_file)} exists")
+        while not self._is_lock_available():
+            self.log.info("Waiting for virtualenv lock to be released")
+            time.sleep(1)
+
+        self.log.info(f"Acquiring the virtualenv lock")
+        self.__acquire_venv_lock()
+        py_bin = self.prepare_virtualenv()
+
+        if not self.is_virtualenv_dir_temporary:
+            self.log.info("Releasing virtualenv lock")
+            self.__release_venv_lock()
+
+        return py_bin
+
+    @property
+    def __lock_file(self) -> Path:
+        return Path(f"{self.virtualenv_dir}/LOCK")
+
+    @property
+    def _pid(self) -> int:
+        return os.getpid()
+
+    # @depends_on_virtualenv_dir
+    def _is_lock_available(self) -> bool:
+        if self.__lock_file.is_file():
+            with open(self.__lock_file) as lf:
+                pid = int(lf.read())
+
+                self.log.info(f"Checking for running process with PID {pid}")
+                try:
+                    _process_running = psutil.Process(pid).is_running()
```

Comment:
Given that we're using subprocesses as well, we must check not only if the pid is running but also if its parent (`os. getppid()`) and potentially the children as well, before releasing the lock

Reply:
Since this is being created during the task execution, in the worker node, I don't see a scenario where the parent process would be creating the lock - I think we have a good starting point and we can introduce additional logic if needed in future.